src/core/Plasma.cu

```diff
@@ -0,0 +1,1161 @@
+    err = MemoryCopy(t, pd->d_Hz, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, MAGNETIC_FIELD_LABEL + Z_LABEL, UNITS_MAGNETIC_FIELD, DESC_MAGNETIC_FIELD + MAGNETIC_FIELD_LABEL + Z_LABEL);
+
+    err = MemoryCopy(t, pd->d_Jx, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, CURRENT_FIELD_LABEL + X_LABEL, UNITS_NO, CURRENT + CURRENT_FIELD_LABEL + X_LABEL);
+    err = MemoryCopy(t, pd->d_Jy, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, CURRENT_FIELD_LABEL + Y_LABEL, UNITS_NO, CURRENT + CURRENT_FIELD_LABEL + Y_LABEL);
+    err = MemoryCopy(t, pd->d_Jz, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, CURRENT_FIELD_LABEL + Z_LABEL, UNITS_NO, CURRENT + CURRENT_FIELD_LABEL + Z_LABEL);
+
+    err = MemoryCopy(t, pd->d_Qx, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, MAGNETIC_HALF_STEP_FIELD_LABEL + X_LABEL, UNITS_NO, DESC_HALFSTEP + MAGNETIC_HALF_STEP_FIELD_LABEL + X_LABEL);
+    err = MemoryCopy(t, pd->d_Qy, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, MAGNETIC_HALF_STEP_FIELD_LABEL + Y_LABEL, UNITS_NO, DESC_HALFSTEP + MAGNETIC_HALF_STEP_FIELD_LABEL + Y_LABEL);
+    err = MemoryCopy(t, pd->d_Qz, sizeof(double) * (Nx + 2) * (Ny + 2) * (Nz + 2), DEVICE_TO_HOST);
+    CHECK_ERROR("MEM COPY", err);
+    writeOne3DArray(filename.c_str(), t, MAGNETIC_HALF_STEP_FIELD_LABEL + Z_LABEL, UNITS_NO, DESC_HALFSTEP + MAGNETIC_HALF_STEP_FIELD_LABEL + Z_LABEL);
+
+    delete[] t;
+}
+
+/**
+ * @param step {int} - step number
+ */
+void Plasma::Step(int step) {
+    int Nx = pd->nx, Ny = pd->ny, Nz = pd->nz;
+
+    ComputeField_FirstHalfStep(step);
+
+    PushParticles(step);
+
+    cout << "push ended" << endl;
+
+    ComputeField_SecondHalfStep(step);
+    cout << "field computed-2" << endl;
+
+    sumMPI((Nx + 2) * (Ny + 2) * (Nz + 2), pd->d_Jx, pd->d_Jy, pd->d_Jz);
+
+    Diagnose(step);
+}
+
+int Plasma::Compute() {
+    cout << "----------------------------------------------------------- " << endl;
+    size_t m_free, m_total;
+
+    cudaMemGetInfo(&m_free, &m_total);
+
+    if (pd->st <= 0 || pd->ts <= 0) {
+        cout << "Invalid computation parameters values!" << endl;
+        return 1;
+    }
+
+    for (int step = pd->st; step <= pd->ts; step++) {
+        memory_monitor("before step", step);
+
+        Step(step);
+
+        // save file
+        if (pd->startSave >= 0 && pd->saveStep >= 0) {
```

Comment:
does it really mean "start saving from the step `startSave` and do it every `saveStep` steps"?